### PR TITLE
Add cmesh example triangulated spherical surface [1/n]

### PR DESCRIPTION
**Pull request by jmark**
**_Describe your changes here:_**

Title says it all. This is a follow-PR to https://github.com/DLR-AMR/t8code/pull/753.

In addition to that, I extended the vtk output by a `diameter` field which is reveals how large the difference
between element sizes is. Generally, it is the goal to have elements of equal sizes in case of uniform refinement.
grafik

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [x] The reviewer executed the new code features at least once and checked the results manually

- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] New source/header files are properly added to the Makefiles
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [x] The code is covered in an existing or new test case using Google Test

#### Github action

- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [x] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [x] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 1]
```diff
@@ -2738,3 +2738,73 @@ t8_cmesh_new_squared_disk (const double radius, sc_MPI_Comm comm)
...
+  double all_verts[ntrees * T8_ECLASS_MAX_CORNERS * T8_ECLASS_MAX_DIM];
+  t8_eclass_t all_eclasses[ntrees];
+
+  /* Defitition of the tree class. */
+  for (int itree = 0; itree < ntrees; itree++) {
+    t8_cmesh_set_tree_class (cmesh, itree, T8_ECLASS_TRIANGLE);
+    all_eclasses[itree] = T8_ECLASS_TRIANGLE;
+  }
+
+  const double r = radius; /* Only for readability. */
```
Comment (Davknapp) [inline src/t8_cmesh/t8_cmesh_examples.c:2766]:
```suggestion
```
We should use explaining variable names. radius is actually a good name.

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 2]
```diff
@@ -2738,3 +2738,73 @@ t8_cmesh_new_squared_disk (const double radius, sc_MPI_Comm comm)
...
+
+  double vertices_top[3][3] = { { r, 0.0, 0.0 }, { 0.0, r, 0.0 }, { 0.0, 0.0, r } };
```
Comment (Davknapp) [inline src/t8_cmesh/t8_cmesh_examples.c:2768]:
```suggestion
  double vertices_top[3][3] = { { radius, 0.0, 0.0 }, { 0.0, radius, 0.0 }, { 0.0, 0.0, radius } };
```

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 3]
```diff
@@ -2738,3 +2738,73 @@ t8_cmesh_new_squared_disk (const double radius, sc_MPI_Comm comm)
...
+
+  double vertices_bot[3][3] = { { r, 0.0, 0.0 }, { 0.0, r, 0.0 }, { 0.0, 0.0, -r } };
```
Comment (Davknapp) [inline src/t8_cmesh/t8_cmesh_examples.c:2770]:
```suggestion
  double vertices_bot[3][3] = { { radius, 0.0, 0.0 }, { 0.0, radius, 0.0 }, { 0.0, 0.0, -radius } };
```

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 4]
```diff
@@ -113,6 +113,224 @@ t8_geometry_squared_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreei
...
+                                                              double *out_coords) const
+{
+  double n[3]; /* Normal vector of the current triangle. */
+  double r[3]; /* Radial vector through one of triangle's corners. */
+  double p[3]; /* Position vector on the triangle plane. */
+
+  t8_locidx_t ltreeid = t8_cmesh_get_local_id (cmesh, gtreeid);
+  double *tree_vertices = t8_cmesh_get_tree_vertices (cmesh, ltreeid);
+
+  n[0] = tree_vertices[0] + tree_vertices[3] + tree_vertices[6];
```
Comment (Davknapp) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:135]: We have a functions for such vector-operations. Please use them.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 5]
```diff
@@ -113,6 +113,224 @@ t8_geometry_squared_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreei
...
+ * \param [in]  gtreeid    The global tree (of the cmesh) in which the reference point is.
+ * \param [in]  ref_coords  Array of \a dimension many entries, specifying a point in [0,1]^dimension.
+ * \param [out] out_coords  The mapped coordinates in physical space of \a ref_coords.
+ */
+void
+t8_geometry_triangulated_spherical_surface::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreeid,
+                                                              const double *ref_coords, const size_t num_coords,
+                                                              double *out_coords) const
+{
+  double n[3]; /* Normal vector of the current triangle. */
```
Comment (Davknapp) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:128]: Please use more explaining names, like normal, radial, position for example.

**Review by Davknapp — CHANGES_REQUESTED** [review]
Loogs good in general, please address my comments.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx [hunk 6]
```diff
@@ -113,6 +113,224 @@ t8_geometry_squared_disk::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreei
...
+  {
+    /* Normalize vector `r`. */
+    const double norm = sqrt (r[0] * r[0] + r[1] * r[1] + r[2] * r[2]);
+    r[0] = r[0] / norm;
+    r[1] = r[1] / norm;
+    r[2] = r[2] / norm;
+  }
+
+  /* Init output coordinates with zeros. */
+  for (size_t i_coord = 0; i_coord < num_coords; i_coord++) {
```
Comment (Davknapp) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx:160]: I don't think you need to zero-initialize, as in line 221 - 223 you can directly set the out_coords to R*p[0]

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx (hunk 4 above)
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx]: Done.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx (hunk 5 above)
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx]: Done.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx (hunk 6 above)
Comment (jmark) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.cxx]: Done.

**Comment by jmark** [comment]
@Davknapp I refactored the code. I realized that I did some unnecessary computations which I removed in this version. The code also becomes simpler that way.

**Review by Davknapp — CHANGES_REQUESTED** [review]

**Review by Davknapp — APPROVED** [review]

**Pull request merged**